File: WebClient SDK/src/default/ui/krProcess/stageHandlers/universalTaskStageTypeUIHandler.ts
import { CardRow, CardRowState } from 'tessa/cards';
import { DefaultCardTypes, sectionName } from 'tessa/workflow';
import { DotNetType, Guid } from 'tessa/platform';
import { GridRowAction, GridRowEventArgs, GridViewModel } from 'tessa/ui/cards/controls';
import {
  IKrStageTypeUIHandlerContext,
  KrStageTypeUIHandler,
  StageTypeHandlerDescriptor,
  universalTaskDescriptor
} from 'tessa/workflow/krProcess';
import {
  IValidationResultBuilder,
  ValidationResultBuilder,
  ValidationResultType
} from 'tessa/platform/validation';

import { ArrayStorage } from 'tessa/platform/storage';
import { showNotEmpty } from 'tessa/ui';

/**
 * UI обработчик типа этапа {@link universalTaskDescriptor}.
 */
export class UniversalTaskStageTypeUIHandler extends KrStageTypeUIHandler {
  //#region fields

  private static readonly _krUniversalTaskOptionsSettingsVirtualSynthetic = sectionName(
    'KrUniversalTaskOptionsSettingsVirtual'
  );

  //#endregion

  //#region base overrides

  public descriptors(): StageTypeHandlerDescriptor[] {
    return [universalTaskDescriptor];
  }

  public async initialize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    const grid = context.settingsForms
      .find(i => i.name === DefaultCardTypes.KrUniversalTaskStageTypeSettingsTypeName)
      ?.blocks.find(i => i.name === 'MainInfo')
      ?.controls.find(i => i.name === 'CompletionOptions') as GridViewModel;

    if (grid) {
      grid.rowInvoked.add(this.rowInvoked);
      grid.rowEditorClosing.add(this.rowClosing);
    }
  }

  public async finalize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    const grid = context.settingsForms
      .find(i => i.name === DefaultCardTypes.KrUniversalTaskStageTypeSettingsTypeName)
      ?.blocks.find(i => i.name === 'MainInfo')
      ?.controls.find(i => i.name === 'CompletionOptions') as GridViewModel;

    if (grid) {
      grid.rowInvoked.remove(this.rowInvoked);
      grid.rowEditorClosing.remove(this.rowClosing);
    }
  }

  //#endregion

  //#region private methods

  private rowInvoked(args: GridRowEventArgs): void {
    if (args.action === GridRowAction.Inserted) {
      args.row.set('OptionID', Guid.newGuid(), DotNetType.Guid);
    }
  }

  private async rowClosing(args: GridRowEventArgs): Promise<void> {
    const row = args.row;
    let validationResult: IValidationResultBuilder | undefined;

    const optionId = row.get('OptionID');
    if (optionId) {
      const rows = args.cardModel.card.sections.get(
        UniversalTaskStageTypeUIHandler._krUniversalTaskOptionsSettingsVirtualSynthetic
      )?.rows;

      if (
        rows &&
        UniversalTaskStageTypeUIHandler.checkDuplicatesOptionId(rows, row.rowId, optionId)
      ) {
        validationResult ??= new ValidationResultBuilder();
        validationResult.add(
          ValidationResultType.Error,
          '$KrProcess_UniversalTask_CompletionOptionIDNotUnique'
        );
        args.cancel = true;
      }
    } else {
      validationResult ??= new ValidationResultBuilder();
      validationResult.add(
        ValidationResultType.Error,
        '$KrProcess_UniversalTask_CompletionOptionIDEmpty'
      );
      args.cancel = true;
    }

    if (!row.tryGet('Caption')) {
      validationResult ??= new ValidationResultBuilder();
      validationResult.add(
        ValidationResultType.Error,
        '$KrProcess_UniversalTask_CompletionOptionCaptionEmpty'
      );
      args.cancel = true;
    }

    if (validationResult) {
      await showNotEmpty(validationResult.build());
    }
  }

  private static checkDuplicatesOptionId(
    rows: ArrayStorage<CardRow>,
    rowId: guid,
    optionId: guid
  ): boolean {
    for (const row of rows) {
      if (Guid.equals(row.rowId, rowId) || row.state === CardRowState.Deleted) {
        continue;
      }

      const iOptionId = row.tryGet('OptionID');

      if (!iOptionId) {
        continue;
      }

      if (Guid.equals(optionId, iOptionId)) {
        return true;
      }
    }

    return false;
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/StageHandlers/UniversalTaskStageTypeUIHandler.cs
#nullable enable

using System;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards.Controls;
using static Tessa.Extensions.Default.Shared.Workflow.KrProcess.KrConstants;

namespace Tessa.Extensions.Default.Client.UI.KrProcess.StageHandlers
{
    /// <summary>
    /// UI обработчик типа этапа <see cref="StageTypeDescriptors.UniversalTaskDescriptor"/>.
    /// </summary>
    public class UniversalTaskStageTypeUIHandler :
        StageTypeUIHandlerBase
    {
        #region Base Overrides

        /// <inheritdoc />
        public override Task Initialize(
            IKrStageTypeUIHandlerContext context)
        {
            if (context.SettingsForms.FirstOrDefault(static i => i.Name == DefaultCardTypes.KrUniversalTaskStageTypeSettingsTypeName)
                ?.Blocks.FirstOrDefault(static i => i.Name == "MainInfo")
                ?.Controls.FirstOrDefault(static i => i.Name == "CompletionOptions") is GridViewModel grid)
            {
                grid.RowInvoked += RowInvoked;
                grid.RowEditorClosing += RowClosing;
            }

            return Task.CompletedTask;
        }

        /// <inheritdoc />
        public override Task Finalize(
            IKrStageTypeUIHandlerContext context)
        {
            if (context.SettingsForms.FirstOrDefault(static i => i.Name == DefaultCardTypes.KrUniversalTaskStageTypeSettingsTypeName)
                ?.Blocks.FirstOrDefault(static i => i.Name == "MainInfo")
                ?.Controls.FirstOrDefault(static i => i.Name == "CompletionOptions") is GridViewModel grid)
            {
                grid.RowInvoked -= RowInvoked;
                grid.RowEditorClosing -= RowClosing;
            }

            return Task.CompletedTask;
        }

        #endregion

        #region Private Methods

        private static void RowInvoked(object? sender, GridRowEventArgs e)
        {
            if (e.Action == GridRowAction.Inserted)
            {
                e.Row.Fields[KrUniversalTaskOptionsSettingsVirtual.OptionID] = Guid.NewGuid();
            }
        }

        private static void RowClosing(
            object? sender,
            GridRowEventArgs e)
        {
            var row = e.Row;
            IValidationResultBuilder? validationResult = null;

            var optionID = row.TryGet<Guid?>(KrUniversalTaskOptionsSettingsVirtual.OptionID);

            if (optionID.HasValue)
            {
                var rows = e.CardModel.Card.Sections[KrUniversalTaskOptionsSettingsVirtual.Synthetic].Rows;

                if (CheckDuplicatesOptionID(rows, row.RowID, optionID.Value))
                {
                    validationResult ??= new ValidationResultBuilder();
                    validationResult.AddError(nameof(UniversalTaskStageTypeUIHandler), "$KrProcess_UniversalTask_CompletionOptionIDNotUnique");
                    e.Cancel = true;
                }
            }
            else
            {
                validationResult ??= new ValidationResultBuilder();
                validationResult.AddError(nameof(UniversalTaskStageTypeUIHandler), "$KrProcess_UniversalTask_CompletionOptionIDEmpty");
                e.Cancel = true;
            }

            if (string.IsNullOrEmpty(row.TryGet<string>(KrUniversalTaskOptionsSettingsVirtual.Caption)))
            {
                validationResult ??= new ValidationResultBuilder();
                validationResult.AddError(nameof(UniversalTaskStageTypeUIHandler), "$KrProcess_UniversalTask_CompletionOptionCaptionEmpty");
                e.Cancel = true;
            }

            if (validationResult is not null)
            {
                TessaDialog.ShowNotEmpty(validationResult);
            }
        }

        private static bool CheckDuplicatesOptionID(
            ListStorage<CardRow> rows,
            Guid rowID,
            Guid optionID)
        {
            foreach (var row in rows)
            {
                if (row.RowID == rowID
                    || row.State == CardRowState.Deleted)
                {
                    continue;
                }

                var iOptionID = row.TryGet<Guid?>(KrUniversalTaskOptionsSettingsVirtual.OptionID);

                if (!iOptionID.HasValue)
                {
                    continue;
                }

                if (optionID == iOptionID)
                {
                    return true;
                }
            }

            return false;
        }

        #endregion
    }
}
